addons/pt/saft/invoices.go

```diff
@@ -55,9 +65,80 @@ func validateInvoice(inv *bill.Invoice) error {
 			),
 			validation.Skip,
 		),
+		validation.Field(&inv.Payment,
+			validation.By(validatePaymentDetails(inv)),
+			validation.Skip,
+		),
+		validation.Field(&inv.Totals,
+			validation.By(validateTotals(inv)),
+			validation.Skip,
+		),
+		validation.Field(&inv.Preceding,
+			validation.Each(
+				validation.By(validatePreceding(inv)),
+				validation.Skip,
+			),
+			validation.Skip,
+		),
 	)
 }
 
+func validatePaymentDetails(inv *bill.Invoice) validation.RuleFunc {
+	return func(val any) error {
+		pay, _ := val.(*bill.PaymentDetails)
+		if pay == nil {
+			return nil
+		}
+
+		return validation.ValidateStruct(pay,
+			validation.Field(&pay.Advances,
+				validation.Each(
+					validation.By(validateAdvance(inv)),
+					validation.Skip,
+				),
+				validation.Skip,
+			),
+		)
+	}
+}
+
+func validateAdvance(inv *bill.Invoice) validation.RuleFunc {
+	return func(val any) error {
+		adv, _ := val.(*pay.Advance)
+		if adv == nil {
+			return nil
+		}
+
+		return validation.ValidateStruct(adv,
+			validation.Field(&adv.Date,
+				validation.Required,
+				validation.In(inv.IssueDate).Error("must be the same as the invoice issue date"),
+				validation.Skip,
+			),
+		)
+	}
+}
+
+func validateTotals(inv *bill.Invoice) validation.RuleFunc {
+	return func(val any) error {
+		tot, _ := val.(*bill.Totals)
+		if tot == nil {
+			return nil
+		}
+
+		return validation.ValidateStruct(tot,
+			validation.Field(&tot.Due,
+				validation.When(isInvoiceReceipt(inv), num.Max(num.AmountZero)),
```

Comment:
Should this not be an Equals zero?

Reply:
yes, but I don't think we have a `num.Equals` or `validation.Equals` validation at the moment. Based on what we have at the moment, it was either `number.Max` or `validation.In`. Would you prefer `validation.In` (if it works with amounts) or should we add a `num.Equals` validation function?